src/ralph/api/auth/oidc.py

```diff
@@ -0,0 +1,147 @@
+    except requests.exceptions.RequestException as exc:
+        logger.error(
+            "Unable to discover the authentication server configuration: %s", exc
+        )
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Could not validate credentials",
+            headers={"WWW-Authenticate": "Bearer"},
+        ) from exc
+
+
+@lru_cache()
+def get_public_keys(jwks_uri: AnyUrl) -> dict:
+    """Retrieve the public keys used by the provider server for signing."""
+    try:
+        response = requests.get(jwks_uri, timeout=5)
+        response.raise_for_status()
+        return response.json()
+    except requests.exceptions.RequestException as exc:
+        logger.error(
+            (
+                "Unable to retrieve the public keys used by the provider server"
+                "for signing: %s"
+            ),
+            exc,
+        )
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Could not validate credentials",
+            headers={"WWW-Authenticate": "Bearer"},
+        ) from exc
+
+
+def get_authenticated_user(
+    auth_header: Union[str, None] = Depends(oauth2_scheme)
+) -> AuthenticatedUser:
+    """Decode and validate OpenId Connect ID token against issuer in config.
+
+    Args:
+        token (str): Base64 encoded OIDC Token. This is invoked behind the
+            scenes by Depends.
+
+    Return:
+        AuthenticatedUser (AuthenticatedUser)
+
+    Raises:
+        HTTPException
+    """
+    if auth_header is None or "Bearer" not in auth_header:
+        logger.error("The OpenID Connect authentication mode requires a Bearer token")
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Could not validate credentials",
+            headers={"WWW-Authenticate": "Bearer"},
+        )
+
+    id_token = auth_header.split(" ")[-1]
+    provider_config = discover_provider(settings.RUNSERVER_AUTH_OIDC_ISSUER_URI)
+    key = get_public_keys(provider_config["jwks_uri"])
+    algorithms = provider_config["id_token_signing_alg_values_supported"]
+    audience = settings.RUNSERVER_AUTH_OIDC_AUDIENCE
+    options = {
+        "verify_signature": True,
+        "verify_aud": bool(audience),
+        "verify_exp": True,
+    }
+    try:
+        decoded_token = jwt.decode(
+            token=id_token,
+            key=key,
+            algorithms=algorithms,
+            options=options,
+            audience=audience,
+        )
+    except (ExpiredSignatureError, JWTError, JWTClaimsError) as exc:
+        logger.error("Unable to decode the ID token: %s", exc)
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Could not validate credentials",
+            headers={"WWW-Authenticate": "Bearer"},
+        ) from exc
+
+    id_token = IDToken.parse_obj(decoded_token)
+
+    return AuthenticatedUser(
```

Comment:
I think `AuthenticatedUser` should be refactored to make a distinction between the various IFI available. It would make more sense to use the `openid` IFI here. Perhaps we could add an `ifi`  or `agent` field to the class ?

Reply:
Letting as it is for now, but I take note to do this for v4.0!